### ISSUE TITLE
 Ruby-buildpack-v1.8.19 still contains the file under Creative Commons Attribution-NonCommercial-ShareAlike 3.0 Unported License 

### ISSUE THREAD

**Issue by luisapace** [cloudfoundry/ruby-buildpack#132]
Hi,
I'm referring to CF Deployment 12.2 and ruby-buildpack-v1.8.19.

This ruby-buildpack-v1.8.19 still contain the the file "incorrect_words.yaml" under Creative Commons Attribution-NonCommercial-ShareAlike 3.0 Unported License.

I've just found this file in this new binary:

name: jruby
  version: 9.2.11.1
  uri: https://buildpacks.cloudfoundry.org/dependencies/jruby/jruby_9.2.11.1-ruby-2.5_linux_x64_cflinuxfs3_b596289d.tgz

At the path: jruby-9.2.11.1-0/jruby_9.2.11.1-ruby-2.5_linux_x64_cflinuxfs3_b596289d/lib/target/jruby-stdlib-9.2.11.1.jar/META-INF/jruby.home/lib/ruby/gems/shared/gems/did_you_mean-1.2.0/evaluation/incorrect_words.yaml

I do not have analyzed all the other binaries, because I cannot be sure that they will not change again in the next ruby-buildpack release; I think that my post-checks will not help to solve the issue timely if that file is not eliminated from all the binaries before to release the ruby-buildpack code.

**Comment by kvedurmu** [comment]
@luisapace I'm not seeing this in the 9.2.11.1 binary you linked.
```
wget -O jruby9.2.11 https://buildpacks.cloudfoundry.org/dependencies/jruby/jruby_9.2.11.1-ruby-2.5_linux_x64_cflinuxfs3_b596289d.tgz
--2020-05-20 10:08:32--  https://buildpacks.cloudfoundry.org/dependencies/jruby/jruby_9.2.11.1-ruby-2.5_linux_x64_cflinuxfs3_b596289d.tgz
Resolving buildpacks.cloudfoundry.org (buildpacks.cloudfoundry.org)... 52.84.23.76, 52.84.23.54, 52.84.23.43, ...
Connecting to buildpacks.cloudfoundry.org (buildpacks.cloudfoundry.org)|52.84.23.76|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 63753647 (61M) [binary/octet-stream]
Saving to: ‘jruby9.2.11’

jruby9.2.11                    100%[====================================================>]  60.80M  7.79MB/s    in 9.1s

2020-05-20 10:08:42 (6.71 MB/s) - ‘jruby9.2.11’ saved [63753647/63753647]

|2.6.5| ny-floater-12962 in /tmp/jruby
○ → tar tvf jruby9.2.11 | grep 'incorrect_words'

  |2.6.5| ny-floater-12962 in /tmp/jruby
○ → echo $?
1
```

Am I missing something?

**Comment by luisapace** [comment]
@kvedurmu thanks for quick followup; you can find it unpacking the jruby-stdlib-9.2.11.1.jar, and then in the META-INF/jruby.home/lib/ruby/gems/shared/gems/did_you_mean-1.2.0/evaluation/incorrect_words.yaml

**Comment by kvedurmu** [comment]
Interesting. So it looks like the PRs that were submitted by your team looks for the `incorrect_words.yaml` file in the file path via a generic find. If your team could submit a PR for this we'd be happy to take a look.

**Comment by andrew-edgar** [comment]
@kvedurmu can you tell me whether I can use "zip" in the find?  because that would allow me to add additional logic to be able to delete from embedded jar files without having to extract then and rear them?

As in this case the bad file is inside a jar file that comes from an external source and is not extracted anytime during the build process.

**Comment by luisapace** [comment]
Thanks, we are working on it!

**Comment by luisapace** [comment]
@kvedurmu @andrew-edgar Hi, I'm sorry to write you again, are there any news on this issue?
The legal team is fully stacked on that and this is blocking also the legal work on CF Deployment 13. Let me know If I can do anything to help to proceed. Thanks.

**Comment by andrew-edgar** [comment]
I am waiting for some feedback from @kvedurmu I'm not going to submit a PR if it's just going to be knocked back.  I have no access to a test env to test this so I need some help from the buildpack team here.

**Comment by dumez-k** [comment]
@luisapace @andrew-edgar

Hi, sorry lots of other stuff on our plate, getting back to this now.

As we use zip in other places, this should work in our environments as well. 

In order to format this command in the Runner module, you will want to call run like this, with only three arguments: ``` Runner.run('bash', '-c', '[your entire command]')```

Does this make sense? You need to explicitly call the ```bash``` command, otherwise you would have to separate this into smaller commands.

**Comment by andrew-edgar** [comment]
I have created PR https://github.com/cloudfoundry/buildpacks-ci/pull/70

**Comment by luisapace** [comment]
@dumez-k  @kvedurmu Hi, this is just to highlight Andrew's PR creation (I'm not sure the system sent you the update). Thanks!

**Comment by dumez-k** [comment]
Approved and merged, waiting for this to make it's way through our pipelines, we'll do validation to make sure the files are gone and then we can hopefully close this out.

**Comment by kvedurmu** [comment]
@luisapace @andrew-edgar unfortunately, it's looking like this PR isn't working as expected. After the PR was merged, the team [rebuilt](https://github.com/cloudfoundry/ruby-buildpack/commit/2dc8f49c11204c975bdfc815fa6ab99e5fb53065) Jruby 9.2.11.1 using these changes and it appears that there is still an `incorrect_words.yaml` file in the jar. Here's what I'm seeing:

1. Download latest binary
```
wget -O jruby.tgz https://buildpacks.cloudfoundry.org/dependencies/jruby/jruby_9.2.11.1-ruby-2.5_linux_x64_cflinuxfs3_54b24279.tgz
```
2. Untar it
```
tar xvf jruby.tgz
```
3. Run the command above
```
find . -name "*.jar" -exec grep -l incorrect_words.yaml {} \; | xargs -I {} zip -q -d {} "*incorrect_words.yaml"
```
4. Run the find command again. You can see that the file wasn't removed at the following path:
```
find . -name "*.jar" -exec grep -l incorrect_words.yaml {} \;
./lib/target/jruby-stdlib-9.2.11.1.jar
```

So, it appears that the recursive find command to find 'incorrect_words.yaml' files is working, but the command to remove the files from the zip is not.

**Comment by andrew-edgar** [comment]
Can you please investigate why it didn't work?  I do NOT have an environment where I can perform the same operations.  You need to give us some more information on what to do.  The round trip time of this is not going to work.  Please help us more by showing us the output of the commands and or helping debug this in the environment where it was happening.  This command worked on my MacBook but if it doesn't work on the env you have then we need your help.

**Comment by andrew-edgar** [comment]
I ran this ...

```
wget -O jruby.tgz https://buildpacks.cloudfoundry.org/dependencies/jruby/jruby_9.2.11.1-ruby-2.5_linux_x64_cflinuxfs3_54b24279.tgz

tar xvzf rjuby.tgz

find . -name "*.jar" -exec grep -l incorrect_words.yaml {} \; | xargs
./lib/target/jruby-stdlib-9.2.11.1.jar

find . -name "*.jar" -exec grep -l incorrect_words.yaml {} \; | xargs -I {} zip -q -d {} '*incorrect_words.yaml'

find . -name "*.jar" -exec grep -l incorrect_words.yaml {} \; | xargs

```

So this is working can you show us the output from the build and what is not working there?  What versions of the "zip" command do you have does it not support -d?

**Comment by kvedurmu** [comment]
@andrew-edgar if my earlier comment wasn't clear, the latest jruby 9.2.11.1 binary (https://buildpacks.cloudfoundry.org/dependencies/jruby/jruby_9.2.11.1-ruby-2.5_linux_x64_cflinuxfs3_54b24279.tgz) has been re-built using your latest PR. Our CI containers are running Zip version `3.0`. 

So, I was expecting that when I download the re-built binary and run the find command **without** the `zip -d` command, I should see no output because the file has already been removed. 

When I run the exact commands you mentioned above on my Macbook, it works as expected, so there must be something going on in the [`#strip_incorrect_words_yaml_from_tar`](https://github.com/cloudfoundry/buildpacks-ci/blob/683eca2dff44803b822da374c21543e5387e2861/tasks/build-binary-new/builder.rb#L410) function.

We'll try to kick this job off again and see if we can dig into the root cause and get back to you.

**Comment by andrew-edgar** [comment]
@kvedurmu yeah I understood.  the strip_incorrect_words_yaml_from_tar should be calling that zip -d and since that worked on my MacBook I would have expected that to work in your build pipeline.  

OK I think I "might" know what the issue is.

In the update I put this ... 

```
        Runner.run('bash', '-c', 'find . -name "*.jar" -exec grep -l incorrect_words.yaml {} \; | xargs -I {} zip -q -d {} "*incorrect_words.yaml"')
 ```
But what we really want is ...

```
        Runner.run('bash', '-c', 'find <dir> -name "*.jar" -exec grep -l incorrect_words.yaml {} \; | xargs -I {} zip -q -d {} "*incorrect_words.yaml"')
```

do you know how I can change the find to have the variable "dir" instead of "."

Since this one only has a param for the -c I need somewhat to fill in the "dir" variable to use instead of '.'

**Comment by dumez-k** [comment]
Hi @andrew-edgar, looking into this now and came to the same conclusion. The find was looking in the wrong directory. 

In order to pass in the dir variable, all you should need to do is wrap it in ```${}```. 

The full command will look like: ``` Runner.run('bash', '-c', 'find ${dir} -name "*.jar" -exec grep -l incorrect_words.yaml {} \; | xargs -I {} zip -q -d {} "*incorrect_words.yaml"')```

Before you pr in this change however, I want to validate it on our environment first. Give me a few minutes. I'll let you know if it works!

**Comment by luisapace** [comment]
@kvedurmu and @andrew-edgar first of all, let me thanks you for your work on this! I've a question for @kvedurmu: assuming that you will release a new ruby-buildpack without that file; in the future will you able to check the presence of that file also in the next ruby-builpacks and eliminate it from them if needed without our involvement or will you need other new PRs from us to do it? Thanks

**Comment by kvedurmu** [comment]
@luisapace once the latest changes are made and PR is merged, I'd suggest that we work together and do one final pass of **all** ruby/jruby dependencies and confirm that the file is removed everywhere before we cut a release.

I can talk with our team about adding test coverage around this so that we aren't accidentally adding the file back in future release. Obviously, it's a little tricky to add 100% coverage around this due to the file being in .jar and other tar artifacts.

**Comment by luisapace** [comment]
@kvedurmu I'm not a developer specialist, so maybe my question is stupid: it should be possible to run commands similar to that in PR, but parameterized so they can do a search of that file on the entire buildpack, shouldn't it? thanks again

**Comment by dumez-k** [comment]
@luisapace @andrew-edgar to speed things up I went ahead and put the changes in myself after verifying they worked in our environments.

[HERE](https://buildpacks.cloudfoundry.org/dependencies/jruby/jruby_9.2.11.1-ruby-2.5_linux_x64_cflinuxfs3_95647f9a.tgz) is the latest build.

Could not find the offending file anywhere.

**Comment by andrew-edgar** [comment]
@dumez-k thanks for making the change!  super happy this worked and removed the file.  We appreciate the help for sure.

**Comment by kvedurmu** [comment]
@andrew-edgar @luisapace could one of you validate that the latest re-built binaries on the develop branch look good (https://github.com/cloudfoundry/ruby-buildpack/blob/develop/manifest.yml). Once you do that, we'll work on getting another release out. Thanks!

**Comment by andrew-edgar** [comment]
I took a look at the dependencies and it does look good to me.  thanks

**Comment by luisapace** [comment]
@kvedurmu I've opened the binaries at that manifest.yml and I've found them free of the bad file!!!
waiting for the new release!
 
@andrew-edgar @kvedurmu @dumez-k thanks a lot of all of you!

**Comment by kvedurmu** [comment]
Hi @luisapace @andrew-edgar thanks for confirming! We've just released [v1.8.20](https://github.com/cloudfoundry/ruby-buildpack/releases/tag/v1.8.20) of the ruby buildpack which has the re-built jruby binaries without this file. 

I'm hoping this is the end of incorrect_words.yaml so I'm going to close this out, but feel free to raise another issue if you observe this file anywhere else 😄.
